wagtail/wagtailcore/models.py

```diff
@@ -1316,17 +1322,191 @@ def serve_password_required_response(self, request, form, action_url):
+            return Page.objects.none()
+
+    def get_explorable_ancestors(self, user):
+        """
+        Returns a queryset of all the node's ancestors that the given user is permitted to see in the Explorer.
+        """
+        # Superusers see all the ancestors.
+        if user.is_superuser:
+            return self.get_ancestors()
+
+        # Copied this algorithm from TreeQuerySet.ancestor_of_q, but tweaked it to remove this page's path.
+        ancestor_paths = [
+            self.path[0:pos]
+            for pos in range(0, len(self.path) + 1 - self.steplen, self.steplen)[1:]
+        ]
+
+        cca_path = get_closest_common_ancestor_path(user)
+        if cca_path:
+            # Return only pages with paths that are at least as long as the Closest Common Ancestor's path.
+            explorable_ancestor_paths = [path for path in ancestor_paths if len(path) >= len(cca_path)]
+            return Page.objects.filter(path__in=explorable_ancestor_paths)
+        else:
+            # If the user has no permitted paths, they have no explorable ancestors.
+            return Page.objects.none()
+
+    def is_explorable_root(self, user):
+        # A superuser's explorable root is the Root page.
+        if user.is_superuser:
+            return self.is_root()
+
+        return self.path == get_closest_common_ancestor_path(user)
+
+    def is_on_site(self, site):
+        """
+        Returns True if this Page is decendent from the given Site's root page.
+        """
+        return Page.objects.descendant_of(site.root_page, inclusive=True).filter(pk=self.id).exists()
+
     class Meta:
         verbose_name = _('page')
         verbose_name_plural = _('pages')
 
 
-def get_navigation_menu_items():
+def get_explorable_page_paths(user):
+    """
+    Returns a tuple of lists of treebeard "path" values.
+    The first list is the set of paths which the given user is explicitly permitted to administer (meaning they can
+    administer all children of those paths, too).
+    The second list is the paths which must be shown to the user in order for them to navigate to their permitted Pages
+    in the Explorer. Unlike the paths in the first list, the children of these paths are not to be automatically made
+    visible.
+
+    NOTE: The first element of the second list is always the Closest Common Ancestor of the User's permitted Pages.
+    """
+    # We cache the output of this function because it's very expensive to generate it fresh every time the user loads
+    # an Explorer page. Unfortunately, this does lead to a lot of duplciated data in the cache, as every single user
+    # that has the same set of groups will have the same permitted paths. I'd like to cache by groupset, which would
+    # remove all that duplication, but that would require an extra DB call every time, to get the user's group list.
+    cache_key = EXPLORABLE_PATHS_PREFIX.format(user.pk)
+    permitted_paths, required_ancestors = cache.get(cache_key, ([], []))
```

Comment:
I'll defer to your expertise here about the call not looking very expensive. I'm certainly not an expert in database performance. Plus, it'll definitely be nice to not need all that extra code for manual cache clearing. I'll make this change right away.